quesma/model/search_after.go

```diff
@@ -0,0 +1,184 @@
+	BasicAndFast SearchAfterStrategyType = iota // default for a second
+	Foolproof
+	JustDiscardTheParameter
+	emptySearchAfterTs = int64(-1)
+)
+
+func (s SearchAfterStrategyType) String() string {
+	return []string{"BasicAndFast", "Foolproof", "JustDiscardTheParameter"}[s]
+}
+
+// ---------------------------------------------------------------------------------
+// | Foolproof, but might be a bit slower for gigantic datasets                    |
+// ---------------------------------------------------------------------------------
+
+type (
+	SearchAfterStrategyFoolproof struct {
+		timestampField ColumnRef
+	}
+	searchAfterParsedFoolproof struct {
+		timestampMs int64
+		pkHashes    []string // md5 for now, should be improved to shorten hashes lengths
+	}
+)
+
+// Validate validates the 'searchAfter', which is what came from the request's search_after field.
+func (s SearchAfterStrategyFoolproof) Validate(searchAfter any) error {
+	logger.Debug().Msgf("searchAfter: %v", searchAfter)
+	_, err := s.validateAndParse(searchAfter)
+	return err
+}
+
+func (s SearchAfterStrategyFoolproof) validateAndParse(searchAfter any) (searchAfterParsedFoolproof, error) {
+	empty := searchAfterParsedFoolproof{timestampMs: emptySearchAfterTs}
+	logger.Debug().Msgf("searchAfter: %v", searchAfter)
+	if searchAfter == nil {
+		return empty, nil
+	}
+
+	asArray, ok := searchAfter.([]any)
+	if !ok {
+		return empty, fmt.Errorf("search_after must be an array")
+	}
+	if len(asArray) == 0 {
+		return empty, fmt.Errorf("for foolproof strategy, search_after must have at most one element")
+	}
+
+	var timestampMs int64
+	if shouldBeTimestamp, ok := util.ExtractNumeric64Maybe(asArray[0]); ok {
+		if shouldBeTimestamp >= 0 && util.IsFloat64AnInt64(shouldBeTimestamp) {
+			timestampMs = int64(shouldBeTimestamp)
+		} else {
+			return empty, fmt.Errorf("for foolproof strategy, search_after[0] must be a unix timestamp in milliseconds")
+		}
+	} else {
+		return empty, fmt.Errorf("for foolproof strategy, search_after must be an integer")
+	}
+
+	return searchAfterParsedFoolproof{timestampMs: timestampMs, pkHashes: make([]string, 0)}, nil // TODO add parsing pk hashes
+}
+
+func (s SearchAfterStrategyFoolproof) ApplyStrategyAndTransformQuery(query *Query, searchAfterRaw any) *Query {
+	searchAfter, _ := s.validateAndParse(searchAfterRaw) // we validate during parsing and error there. No need to check here.
+	if searchAfter.timestampMs == emptySearchAfterTs {
+		return query
+	}
+	timestampRangeClause := NewInfixExpr(s.timestampField, "<=", NewFunction("fromUnixTimestamp64Milli", NewLiteral(searchAfter.timestampMs))) // TODO fix this for Hydrolix...
+	logger.Debug().Msgf("search_after: %v, query before: %v", searchAfter, AsString(query.SelectCommand))
+	query.SelectCommand.WhereClause = And([]Expr{query.SelectCommand.WhereClause, timestampRangeClause})
+	query.SelectCommand.Limit += len(searchAfter.pkHashes)
+	logger.Debug().Msgf("query after: %v", AsString(query.SelectCommand))
```

Comment:
There are two debugs, which, even if they are not rendered, they're still quite expensive SQL for every hit. Probably commenting them out would be the easiest solution.